flink-runtime/src/test/java/org/apache/flink/runtime/metrics/MetricReporterSetupTestBase.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.metrics;
+
+import org.apache.flink.metrics.reporter.MetricReporterFactory;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * The base class to verify metric reporter factory.
+ */
+public abstract class MetricReporterSetupTestBase extends TestLogger {
```

Comment:
Let's not introduce inheritance for this. The test could just be a method that is being called with the factory class.

Reply:
If we do not inherit from a base class, we need to add this test method in 5 modules: `flink-metrics-datadog`, `flink-metrics-graphite`, `flink-metrics-influxdb`, `flink-metrics-jmx` and `flink-metrics-statsd` (if we give the background knowledge that Prometheus and slf4j reporter has been used in e2e tests). From my point of view, this would be too redundant. 
I could move the base test class from `flink-runtime` to `flink-metrics-core`  module to get rid of dependency of `flink-runtime` even they're in test scope. What do you think?

Reply:
It's fine to duplicate things. Factory support is optional after all, and having this as a simple opt-in method that external reporter implementations could use without interfering with class structures is useful.

We already see the downsides of inheritance in this PR; the test method must be unnecessarily complex since it must support the Prometheus case where there are multiple reporters, with a signature that for all other reporters doesn't make sense.

Reply:
Okay, I'll update the PR with duplicate test methods to see the final effect.